src/contexts/ProcedureProvider.tsx

```diff
@@ -0,0 +1,62 @@
+import { createContext, ReactNode, useCallback, useContext, useMemo, useState } from "react";
+import { v4 as uuidv4 } from "uuid";
+
+export const categories = ["Economy", "Housing", "Employment", "Justice"] as const;
+
+export interface Component {
+  id: string;
+  name: string;
+  properties: object;
+  required?: string[];
+}
+
+export interface Procedure {
+  name: string;
+  category: typeof categories[number];
+  description: string;
+  components: Component[];
+}
+
+interface Context {
+  procedure: Procedure | null;
+  createProcedure: (procedure: Procedure) => void;
+  addComponent: (component: Omit<Component, "id">) => void;
+  deleteComponentById: (id: string) => void;
+}
+
+const ProcedureContext = createContext<Context>({} as Context);
+export const useProcedures = () => useContext(ProcedureContext);
+
+export const ProcedureProvider = ({ children }: { children: ReactNode }) => {
+  const [procedure, setProcedure] = useState<Procedure | null>(null);
+
+  const createProcedure = useCallback((procedure: Procedure) => {
+    setProcedure({ ...procedure, components: [] });
+  }, []);
+
+  const addComponent = useCallback((component: Omit<Component, "id">) => {
+    const newComponent: Component = { ...component, id: uuidv4() };
+
+    setProcedure((oldProcedure) =>
+      oldProcedure ? { ...oldProcedure, components: [...oldProcedure.components, newComponent] } : null
+    );
+  }, []);
+
+  const deleteComponentById = useCallback((id: string) => {
+    setProcedure((oldProcedure) =>
+      oldProcedure ? { ...oldProcedure, components: oldProcedure.components.filter((c) => c.id !== id) } : null
+    );
+  }, []);
+
+  const context: Context = useMemo(
+    () => ({
+      procedure,
+      createProcedure,
+      addComponent,
+      deleteComponentById,
+    }),
+    [procedure, createProcedure, addComponent, deleteComponentById]
+  );
```

Comment:
Is it useful to memoize a value if the dependencies are all dependencies available in the class?

Please correct me if I'm wrong, but I think the behavior of the code will be the same without memoization here! :

Reply:
Whoops, I believe you're right! Quite an odd habit we were taught for "optimizing" performance!